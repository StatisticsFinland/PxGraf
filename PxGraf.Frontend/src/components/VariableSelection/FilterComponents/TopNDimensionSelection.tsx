import React from 'react';
import { TextField } from '@mui/material';
import { useTranslation } from 'react-i18next';
import styled from 'styled-components';

interface ITopNDimensionSelectionProps {
    numberOfItems: number,
    onNumberChanged: (newValue: number) => void
}

const StyledTextField = styled(TextField)`
    background-color: #fff;
`;

export const TopNDimensionSelection: React.FC<ITopNDimensionSelectionProps> = ({ numberOfItems, onNumberChanged }) => {
    const { t } = useTranslation();

    const handleChange = (event) => {
        const regExp = /^\d+$/;
        if (regExp.test(event.target.value)) {
            onNumberChanged(parseInt(event.target.value));
        }
        else {
            onNumberChanged(0);
        }
    }

    return (
        <StyledTextField label={t("variableSelect.topFilter")}
            defaultValue={numberOfItems?.toString() ?? ""}
            onChange={handleChange}
            fullWidth
            inputProps={{ inputMode: 'numeric', pattern: '[0-9]*' }}
        />
    );
}

export default TopNDimensionSelection;