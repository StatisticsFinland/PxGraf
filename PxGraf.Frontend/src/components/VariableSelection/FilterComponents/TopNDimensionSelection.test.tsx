import React from 'react';
import { fireEvent, render, screen } from "@testing-library/react";

import TopNDimensionSelection from "./TopNDimensionSelection";

jest.mock('react-i18next', () => ({
    ...jest.requireActual('react-i18next'),
    useTranslation: () => {
        return {
            t: (str: string) => str,
            i18n: {
                changeLanguage: () => new Promise(() => null),
            },
        };
    },
}));

const mockChangeFunction = jest.fn();

describe('Rendering test', () => {
    it('renders correctly', () => {
        const { asFragment } = render(<TopNDimensionSelection
            numberOfItems={2}
            onNumberChanged={(_newValue: number) => { }}
        ></TopNDimensionSelection>);
        expect(asFragment()).toMatchSnapshot();
    });
});

describe('Assertion test', () => {
    it('should call onchange with the correct value when value is convertable to number', () => {
        render(<TopNDimensionSelection
            numberOfItems={2}
            onNumberChanged={mockChangeFunction}
        ></TopNDimensionSelection>);
        fireEvent.change(screen.getByLabelText('variableSelect.topFilter'), { target: { value: '5' } });
        expect(mockChangeFunction).toBeCalledWith(5);
    });

    it('should call onchange with 0 when value is not convertable to number', () => {
        render(<TopNDimensionSelection
            numberOfItems={2}
            onNumberChanged={mockChangeFunction}
        ></TopNDimensionSelection>);
        fireEvent.change(screen.getByLabelText('variableSelect.topFilter'), { target: { value: 'eivoikaantaa' } });
        expect(mockChangeFunction).toBeCalledWith(0);
    });
});