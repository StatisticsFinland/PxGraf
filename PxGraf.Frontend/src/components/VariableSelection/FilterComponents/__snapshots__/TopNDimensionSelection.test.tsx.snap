// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering test renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root sc-blHHSb egjXqi css-wb57ya-MuiFormControl-root-MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
      for=":r0:"
      id=":r0:-label"
    >
      variableSelect.topFilter
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
        id=":r0:"
        inputmode="numeric"
        pattern="[0-9]*"
        type="text"
        value="2"
      />
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-14lo706"
        >
          <span>
            variableSelect.topFilter
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</DocumentFragment>
`;
